### DISCUSSION TITLE
Incredibly close to ripgrep with line number hyperlinking - Status survey

### DISCUSSION THREAD

**Discussion by gczajkowskiTT** [dandavison/delta#1898] [General]
Delta is incredible, and I have been using it for years. I'm attempting to have my team of fifty adopt it as their "tool of choice" in our git-centric workflows. I want to make a great impression during the tool "show-off." 

The powers of delta turn exponential when integrated with other super tools (just like your favorite Marvel & Capcom superheroes teams).

I am pulling together all the various integrations to show off the combined superpowers of Delta, Ripgrep/Ugrep, git, Atuin, Powerline-status, Warp/Kitty/Konsole, fzf, and many new rust system utilities. (Surprisingly, very few know about the powerful ripgrep or ugrep searchers)

Pain point seems to be hyperlinking is not functioning at all for me with delta :( across two OSes and four terminals) focusing on `rg -C 2 needle --json | delta --hyperlinks` it looks like this:

<img width="1743" alt="image" src="https://github.com/user-attachments/assets/df7fceaf-1f7d-4c01-afa7-a2766583ed81">

Test http link: `printf '\e]8;;http://example.com\e\\This is a link\e]8;;\e\\\n'`
Test vscode link: `printf '\e]8;;vscode://file//Users/jaminthornsberry/.gitconfig:10\e\\Take me to git config!\e]8;;\e\n'`

Linux RH8
| Super Power | Command | Warp (not functional & not supported)  | Kitty | Konsole | VSCode (not used)
|---|---|---|---|---|---| 
|Test http link| `printf ...`|NS|Yes|Yes|NU| 
| Ripgrep File/Line Linking | `rg -C 2 needle` | NS | File & Line Links: `rg -C 2 needle --hyperlink-format=kitty` | File | NU
| Delta git blame commit hash linking | `git blame` | NS | Yes (with wrapper) | Yes (with wrapper) | NU
| Ripgrep+Delta File Linking | `rg -C 2 needle --json \| delta --hyperlinks` | NS | Yes (`less -rFX`) | Yes (`less -rFX`) | NU
| Ripgrep+Delta Line# Linking | `rg -C 2 needle --json \| delta --hyperlinks` | NS | Yes (`less -rFX`) | No | NU

| Program | Delta | Ripgrep | Warp | Kitty | Konsole | VSCode
|---|---|---|---|---|---|---|
|Versions|0.18.2|14.1.1 (rev 79cbe89deb)| Not functional with RH8 because of GTK requirements | 0.37.0 created by Kovid Goyal|23.04.1|Not used

MacOS
| Super Power | Command | Warp (Run but hyperlinking not supported)  | Kitty | Konsole (Not Installed) | VSCode Terminal
|---|---|---|---|---|---|
|Test http link| `printf ...`|NS|Yes|NI|Yes|
|Test vscode link| `printf ...`|NS|Broken link|NI|Yes|
| Ripgrep File/Line Linking | `rg -C 2 needle...` | NS | YES: `rg --hyperlink-format=kitty` | NI | YES: `rg --hyperlink-format=vscode` 
| Delta git blame commit hash linking | `git blame` | NS | No Link | NA | No Link
| Ripgrep+Delta File Linking | `rg -C 2 needle --json \| delta --hyperlinks` | NS | YES: `rg -C 2 set_env --json \| delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"` | NA | YES: `rg -C 2 set_env --json \| delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"`
| Ripgrep+Delta Line# Linking | `rg -C 2 needle --json \| delta --hyperlinks` | NS | YES: `rg -C 2 set_env --json \| delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"` | NA | YES:  `rg -C 2 set_env --json \| delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"`

| Program | Delta | Ripgrep | Kitty | Konsole | Warp | VSCode|
|---|---|---|---|---|---|---|
|Versions|0.18.2|14.1.1 (rev 79cbe89deb)|0.37.0 created by Kovid Goyal|23.04.1| v0.2024.11.05.08.02.stable_02 |Version: 1.95.2|

**Comment by gczajkowskiTT** [comment]
https://github.com/dandavison/delta/issues/1397 
https://github.com/dandavison/delta/issues/1628

**Comment by dandavison** [comment]
Hi @gczajkowskiTT, let's get to the bottom of this.

As one starting point I can confirm that the following is working for me (`ripgrep 14.1.0, delta 0.18.2`). File paths and line numbers are correctly hyperlinked

```
command rg --json create | delta --hyperlinks --no-gitconfig
```

Here's some example `bat -A` output from that:

```
␛]8;;file:///Users/dan/src/temporalio/sdk-python/scripts/run_bench.py␛\␛[35mscripts/run_bench.py␛[0m␛]8;;␛\·␊
␛]8;;file:///Users/dan/src/temporalio/sdk-python/scripts/run_bench.py␛\␛[32m72␛[0m␛]8;;␛\:␛[38;2;248;248;242m········report_mem_task·␛[38;2;249;38;114m=␛[38;2;248;248;242m·asyncio.␛[48;2;0;96;0mcreate␛[0m␛[38;2;248;248;242m_task(report_mem())␛[0m␊
␊
```

and `cat -A`

```
^[]8;;file:///Users/dan/src/temporalio/sdk-python/scripts/run_bench.py^[\^[[35mscripts/run_bench.py^[[0m^[]8;;^[\ $
^[]8;;file:///Users/dan/src/temporalio/sdk-python/scripts/run_bench.py^[\^[[32m72^[[0m^[]8;;^[\:^[[38;2;248;248;242m        report_mem_task ^[[38;2;249;38;114m=^[[38;2;248;248;242m asyncio.^[[48;2;0;96;0mcreate^[[0m^[[38;2;248;248;242m_task(report_mem())^[[0m$
```

and hexyl
```
┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
│00000000│ 1b 5d 38 3b 3b 66 69 6c ┊ 65 3a 2f 2f 2f 55 73 65 │•]8;;fil┊e:///Use│
│00000010│ 72 73 2f 64 61 6e 2f 73 ┊ 72 63 2f 74 65 6d 70 6f │rs/dan/s┊rc/tempo│
│00000020│ 72 61 6c 69 6f 2f 73 64 ┊ 6b 2d 70 79 74 68 6f 6e │ralio/sd┊k-python│
│00000030│ 2f 73 63 72 69 70 74 73 ┊ 2f 72 75 6e 5f 62 65 6e │/scripts┊/run_ben│
│00000040│ 63 68 2e 70 79 1b 5c 1b ┊ 5b 33 35 6d 73 63 72 69 │ch.py•\•┊[35mscri│
│00000050│ 70 74 73 2f 72 75 6e 5f ┊ 62 65 6e 63 68 2e 70 79 │pts/run_┊bench.py│
│00000060│ 1b 5b 30 6d 1b 5d 38 3b ┊ 3b 1b 5c 20 0a 1b 5d 38 │•[0m•]8;┊;•\ _•]8│
│00000070│ 3b 3b 66 69 6c 65 3a 2f ┊ 2f 2f 55 73 65 72 73 2f │;;file:/┊//Users/│
│00000080│ 64 61 6e 2f 73 72 63 2f ┊ 74 65 6d 70 6f 72 61 6c │dan/src/┊temporal│
│00000090│ 69 6f 2f 73 64 6b 2d 70 ┊ 79 74 68 6f 6e 2f 73 63 │io/sdk-p┊ython/sc│
│000000a0│ 72 69 70 74 73 2f 72 75 ┊ 6e 5f 62 65 6e 63 68 2e │ripts/ru┊n_bench.│
│000000b0│ 70 79 1b 5c 1b 5b 33 32 ┊ 6d 37 32 1b 5b 30 6d 1b │py•\•[32┊m72•[0m•│
│000000c0│ 5d 38 3b 3b 1b 5c 3a 1b ┊ 5b 33 38 3b 32 3b 32 34 │]8;;•\:•┊[38;2;24│
│000000d0│ 38 3b 32 34 38 3b 32 34 ┊ 32 6d 20 20 20 20 20 20 │8;248;24┊2m      │
│000000e0│ 20 20 72 65 70 6f 72 74 ┊ 5f 6d 65 6d 5f 74 61 73 │  report┊_mem_tas│
│000000f0│ 6b 20 1b 5b 33 38 3b 32 ┊ 3b 32 34 39 3b 33 38 3b │k •[38;2┊;249;38;│
│00000100│ 31 31 34 6d 3d 1b 5b 33 ┊ 38 3b 32 3b 32 34 38 3b │114m=•[3┊8;2;248;│
│00000110│ 32 34 38 3b 32 34 32 6d ┊ 20 61 73 79 6e 63 69 6f │248;242m┊ asyncio│
│00000120│ 2e 1b 5b 34 38 3b 32 3b ┊ 30 3b 39 36 3b 30 6d 63 │.•[48;2;┊0;96;0mc│
│00000130│ 72 65 61 74 65 1b 5b 30 ┊ 6d 1b 5b 33 38 3b 32 3b │reate•[0┊m•[38;2;│
│00000140│ 32 34 38 3b 32 34 38 3b ┊ 32 34 32 6d 5f 74 61 73 │248;248;┊242m_tas│
│00000150│ 6b 28 72 65 70 6f 72 74 ┊ 5f 6d 65 6d 28 29 29 1b │k(report┊_mem())•│
│00000160│ 5b 30 6d 0a 0a 1b 5d 38 ┊ 3b 3b 66 69 6c 65 3a 2f │[0m__•]8┊;;file:/│
│00000170│ 2f 2f 55 73 65 72 73 2f ┊ 64 61 6e 2f 73 72 63 2f │//Users/┊dan/src/│
│00000180│ 74 65 6d 70 6f 72 61 6c ┊ 69 6f 2f 73 64 6b 2d 70 │temporal┊io/sdk-p│
│00000190│ 79 74 68 6f 6e 2f 74 65 ┊ 6d 70 6f 72 61 6c 69 6f │ython/te┊mporalio│
│000001a0│ 2f 77 6f 72 6b 65 72 2f ┊ 5f 77 6f 72 6b 66 6c 6f │/worker/┊_workflo│
│000001b0│ 77 2e 70 79 1b 5c 1b 5b ┊ 33 35 6d 74 65 6d 70 6f │w.py•\•[┊35mtempo│
│000001c0│ 72 61 6c 69 6f 2f 77 6f ┊ 72 6b 65 72 2f 5f 77 6f │ralio/wo┊rker/_wo│
│000001d0│ 72 6b 66 6c 6f 77 2e 70 ┊ 79 1b 5b 30 6d 1b 5d 38 │rkflow.p┊y•[0m•]8│
│000001e0│ 3b 3b 1b 5c 20 0a 1b 5d ┊ 38 3b 3b 66 69 6c 65 3a │;;•\ _•]┊8;;file:│
│000001f0│ 2f 2f 2f 55 73 65 72 73 ┊ 2f 64 61 6e 2f 73 72 63 │///Users┊/dan/src│
│00000200│ 2f 74 65 6d 70 6f 72 61 ┊ 6c 69 6f 2f 73 64 6b 2d │/tempora┊lio/sdk-│
│00000210│ 70 79 74 68 6f 6e 2f 74 ┊ 65 6d 70 6f 72 61 6c 69 │python/t┊emporali│
│00000220│ 6f 2f 77 6f 72 6b 65 72 ┊ 2f 5f 77 6f 72 6b 66 6c │o/worker┊/_workfl│
│00000230│ 6f 77 2e 70 79 1b 5c 1b ┊ 5b 33 32 6d 31 35 39 1b │ow.py•\•┊[32m159•│
│00000240│ 5b 30 6d 1b 5d 38 3b 3b ┊ 1b 5c 3a 1b 5b 33 38 3b │[0m•]8;;┊•\:•[38;│
│00000250│ 32 3b 32 34 38 3b 32 34 ┊ 38 3b 32 34 32 6d 20 20 │2;248;24┊8;242m  │
```


(Incidentally, on a separate note, you might be interested in https://github.com/dandavison/delta/pull/1769 -- feel free to leave any feedback there)

**Comment by gczajkowskiTT** [comment]
<img width="1441" alt="image" src="https://github.com/user-attachments/assets/c9c90085-80de-45ee-9d2e-39e94730a05c">

**Reply by gczajkowskiTT** [reply]
In this kitty terminal `rg --hyperlink-format=kitty create` does work.. it's hard to see but the line number is linked

<img width="481" alt="image" src="https://github.com/user-attachments/assets/f97ffa7d-c793-4aa1-932e-4f0bed9801bc">

**Comment by dandavison** [comment]
I've installed kitty `0.37.0` and delta hyperlinks in diff, rg, and blame output are all working correctly (MacOS) from what I've seen so far.

Are you using tmux?

Would you be able to try Alacritty on MacOS?

It feels to me that the reported problems with hyperlinks must be user-environment-specific: I believe that delta is emitting the correct sequence of bytes, as evidenced by the fact that it works for most people.

**Comment by gczajkowskiTT** [comment]
Hi, I'm not using Tmux. I got a lot further since yesterday on MacOS: 

1) Registered all filetypes with vscode with this gem https://stackoverflow.com/questions/43665346/can-somebody-explain-how-to-make-vscode-the-default-editor-on-osx 
2) `rg -C 2 set_env --json | delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"` works in Kitty and VSCode terminals now
3) `git blame` commit links worked for a hot second in VSCode terminal and then broke again

What's missing:
1) Commit links in `git blame`
2) Line links in `git blame` 
3) Linux is still quite broken, will reply to the other issue

Todo:
1) Simplify `rg -C 2 set_env --json | delta --hyperlinks --hyperlinks-file-link-format "vscode://file/{path}:{line}"` to set defaults in .gitconfig and use Zsh function reducing `rg -C 2 set_env --json | delta` to `rg` function

**Reply by gczajkowskiTT** [reply]
@dandavison made a breakthrough after reading lots of threads on this. To get `kitty` file and line links to work, I set DELTA_PAGER to `less -rFX` and added `kitty` open-actions config to launch the editor and an editor wrapper that parses different "file" paths and adjust the editor launch.

`/path/to/file:line`
`/path/to/file:`
`/path/to/file`

For commit hash links, I created a `delta` wrapper script that extracts the `remote.origin.url` and converts URIs of the form git@ to URLS http:// dynamically per clone and executes delta with `--hyperlinks-commit-link-format`. Hard-coding hyperlinks-commit-link-format in ~/.gitconfig is unusable when juggling dozens of repos.

Nice to haves in Delta

1) Delta is adding a tail ":" during `git diff` headers, which I had to strip out manually. I'd prefer it if Delta didn't add it.
2) Delta automatically converted git@ and git+http@ URIs to http:// URLs for the commit links.
